plugins/checksum/checksum.c

```diff
@@ -22,244 +22,431 @@
+  if (checksum == NULL) {
+    hexchat_commandf(ph, "Checksum: Failed to malloc needed memory\n");
+    return HEXCHAT_EAT_NONE;
+  }
+
+  /* Print in the privmsg tab of the sender */
+  hexchat_set_context(ph, hexchat_find_context(ph, NULL, word[3]));
+
+  if (hexchat_get_prefs(ph, "dcc_completed_dir", &dcc_completed_dir, NULL) ==
+          1 &&
+      dcc_completed_dir[0] != '\0') {
+    filename = g_build_filename(dcc_completed_dir, word[1], NULL);
+  } else {
+    filename = g_strdup(word[2]);
+  }
+
+  dccrecv_async_cb_data *user_data = malloc(sizeof(dccrecv_async_cb_data));
+  if (user_data == NULL) {
+    hexchat_printf(ph, "Checksum: Failed to malloc needed memory\n");
+    free(checksum);
+    return HEXCHAT_EAT_NONE;
+  }
+
+  user_data->checksum = checksum;
+
+  char *word_1 = malloc(strlen(word[1]) + 1);
+  if (word_1 == NULL) {
+    hexchat_printf(ph, "Checksum: Failed to malloc needed memory\n");
+    free(checksum);
+    free(user_data);
+    return HEXCHAT_EAT_NONE;
+  }
+
+  strcpy(word_1, word[1]);
+
+  user_data->word_1 = word_1;
+  user_data->filename = filename;
+
+  sha256_from_file_async(filename, checksum, dccrecv_async_cb,
+                         (void *)user_data);
+
+  return HEXCHAT_EAT_NONE;
+}
 
-	/* Print in the privmsg tab of the receiver */
-	hexchat_set_context (ph, hexchat_find_context (ph, NULL, word[3]));
+typedef struct {
+  char *checksum;
+  char *word_1;
+  char *word_2;
+} dccoffer_async_cb_data;
 
-	if (sha256_from_file (word[3], checksum))
-	{
-		hexchat_commandf (ph, "quote PRIVMSG %s :SHA-256 checksum for %s (remote): %s", word[2], word[1], checksum);
-	}
+void dccoffer_async_cb(gboolean result, gpointer user_data) {
+
+  dccoffer_async_cb_data *bound_vars = (void *)user_data;
+
+  if (result) {
+    hexchat_commandf(
+        ph, "quote PRIVMSG %s :SHA-256 checksum for %s (remote): %s",
+        bound_vars->word_2, bound_vars->word_1, bound_vars->checksum);
+  }
+
+  free(bound_vars->word_2);
+  free(bound_vars->word_1);
+  free(bound_vars->checksum);
+
+  free(bound_vars);
+}
 
-	return HEXCHAT_EAT_NONE;
+static int dccoffer_cb(char *word[], void *userdata) {
+
+  /* Print in the privmsg tab of the receiver */
+  hexchat_set_context(ph, hexchat_find_context(ph, NULL, word[3]));
+
+  char *checksum = malloc(SHA256_BUFFER_LENGTH);
+  if (checksum == NULL) {
+    hexchat_commandf(ph, "Checksum: Failed to malloc needed memory\n");
+    return HEXCHAT_EAT_NONE;
+  }
+
+  dccoffer_async_cb_data *user_data = malloc(sizeof(dccoffer_async_cb_data));
+  if (user_data == NULL) {
+    hexchat_printf(ph, "Checksum: Failed to malloc needed memory\n");
+    free(checksum);
+    return HEXCHAT_EAT_NONE;
+  }
+
+  user_data->checksum = checksum;
+
+  char *word_1 = malloc(strlen(word[1]) + 1);
+  if (word_1 == NULL) {
+    hexchat_printf(ph, "Checksum: Failed to malloc needed memory\n");
+    free(checksum);
+    free(user_data);
+    return HEXCHAT_EAT_NONE;
+  }
+
+  char *word_2 = malloc(strlen(word[2]) + 1);
```

Comment:
Replace all `malloc` in this file.

Use `g_strdup()` to copy strings. `g_new()` to allocate in general.

Reply:
ok, and sorry, I didn't notice the file was formatted, but autoformatters format automatically I guess ;(